src/python/pants/core/target_types.py

```diff
@@ -206,5 +216,142 @@ class GenericTarget(Target):
     core_fields = (*COMMON_TARGET_FIELDS, Dependencies)
 
 
+# -----------------------------------------------------------------------------------------------
+# `archive` target
+# -----------------------------------------------------------------------------------------------
+
+# TODO(#10888): Teach project introspection goals that this is a special type of the `Dependencies`
+#  field.
+class ArchivePackages(StringSequenceField):
+    """Addresses to any targets that can be built with `./pants package`.
+
+    Pants will build the assets as if you had run `./pants package`. It will include the
+    results in your archive using the same name they would normally have, but without the `dist/`
+    prefix.
```

Comment:
I could see this being an issue, if you want the archive to have a different path. It could especially be an issue once we remove the "legacy paths" of only `target_name.ext`, in favor of `path.to.parent/target_name.ext`. You might not want that parent directory to be created.

cc @benjyw if we either a) want to permanently keep the legacy behavior as an option, or b) add a mechanism to rename.

I could also see this being an issue once we convert `setup-py` to use `package`, meaning it's now eligible to be included here. The name for `setup-py` includes the version in its output. That means that your `archive`'s paths will not be safe. We may need a mechanism to rewrite it to something stable.

Reply:
In cases where this might be an issue, we can always require an explicit output file name on the dependency target.

`jvm_binary` and `python_binary` used to have a `basename` field which became the root of the output filename.

Reply:
Yeah, I don't think we want to permanently keep the legacy behavior. We can use the fully-qualified path by default, and allow it to be overridden as @stuhood suggests.

Reply:
I'm not totally following the proposal. We would be adding a new field on each package type called something like `output_path`, right? You can set any arbitrary path for a `python_binary`, `archive`, etc. Right?

If undefined, we default to the `path.to.parent/target_name.ext`.

If this is correct, I can add that in a fast follow-up. I think it's a good suggestion.

Reply:
Yes, I think that sounds right.

Reply:
Just the filename, rather than the whole path, I think? Almost none of the paths that are specified on targets are absolute (other than target addresses), so it would be a bit odd if it was the whole path. It would also enable collisions in dist fairly easily.

Reply:
Oh this conversation got split into two threads, oops. See https://github.com/pantsbuild/pants/pull/10895#discussion_r499048866 for why I think it's really important that we allow you to set the entire path, rather than only the final filename.